File: brainfuck/brainfuck.cs
using System;
using System.IO;
using System.Linq;
using System.Collections.Generic;
using System.Text;

namespace Test
{
    public class Tape
    {
        int pos;
        List<int> tape;

        public Tape()
        {
            pos = 0;
            tape = new List<int>(new int[]{0});
        }

        public int Get() { return tape[pos]; }
        public void Inc() { tape[pos]++; }
        public void Dec() { tape[pos]--; }
        public void Advance() { pos++; if (tape.Count <= pos) tape.Add(0); }
        public void Devance() { if (pos > 0) { pos--; } }
    }

    public struct Op
    {
        public char ch;
        public int jump;
        public Op(char ch1, int jump1)
        {
            ch = ch1;
            jump = jump1;
        }
    }

    class Program
    {
        List<Op> code;

        Program(string text)
        {
            Stack<int> leftstack = new Stack<int>();
            code = new List<Op>();

            for (int i = 0; i < text.Length; i++) {
                if ("[]<>+-,.".IndexOf(text[i]) != -1) {
                    Op op = new Op(text[i], 0);
                    code.Add(op);
                }
            }
            for (int pc = 0; pc < code.Count; pc++) {
                char c = code[pc].ch;
                if (c == '[') leftstack.Push(pc);
                else
                {
                    if (c == ']' && leftstack.Count != 0)
                    {
                        int left = leftstack.Pop();
                        int right = pc;
                        code[left] = new Op(code[left].ch, right);
                        code[right] = new Op(c, left);
                    }
                }
            }
        }

        void Run()
        {
            Tape tape = new Tape();
            for (int pc = 0; pc < code.Count; ++pc)
            {
                Op op = code[pc];
                switch (op.ch)
                {
                case '+':
                    tape.Inc();
                    break;
                case '-':
                    tape.Dec();
                    break;
                case '>':
                    tape.Advance();
                    break;
                case '<':
                    tape.Devance();
                    break;
                case '[':
                    if (tape.Get() == 0) pc = op.jump;
                    break;
                case ']':
                    if (tape.Get() != 0) pc = op.jump;
                    break;
                case '.':
                    Console.Write((char)tape.Get());
                    break;
                }
            }
        }

        static void Main(string[] args)
        {
            string text = File.ReadAllText(args[0]);
            var p = new Program(text);
            p.Run();
        }
    }
}

File: brainfuck/brainfuck.scala
class Tape() {
  private var tape = Array(0)
  private var pos  = 0

  def get = tape(pos)
  def inc = tape(pos) += 1
  def dec = tape(pos) -= 1
  def advance = { pos += 1; if (tape.length <= pos) { tape :+= 0 } }
  def devance = { if (pos > 0) { pos -= 1 } }
}

class Program(text: String) {
  var code = Array.empty[Char]
  var bracket_map = collection.mutable.HashMap.empty[Int, Int]

  parse_code(text)

  def parse_code(text: String) {
    val leftstack = collection.mutable.ArrayStack.empty[Int]
    var pc = 0
    for (ch <- text)
      ch match {

        case '.' | ',' | '+' | '-' | '<' | '>' =>
          code :+= ch
          pc += 1

        case '[' =>
          leftstack push pc
          code :+= ch
          pc += 1

        case ']' =>
          if (leftstack.nonEmpty) {
            val left = leftstack.pop()
            val right = pc
            bracket_map.update(left, right)
            bracket_map.update(right, left)
          }
          code :+= ch
          pc += 1

        case _ =>
      }
  }

  def run = {
    var pc = 0
    var tape = new Tape()
    while ( pc < code.length ) {
      code(pc) match {
        case '>' => tape.advance
        case '<' => tape.devance
        case '+' => tape.inc
        case '-' => tape.dec
        case '[' => if (tape.get == 0) pc = bracket_map(pc)
        case ']' => if (tape.get != 0) pc = bracket_map(pc)
        case '.' => print(tape.get.toChar)
      }
      pc += 1
    }
  }
}

object BrainFuck {
  def time(f: => Any) = {
    val s = System.nanoTime
    val ret = f
    println("time: "+(System.nanoTime-s)/1e9+"s")
    ret
  }

  def main(args: Array[String]): Unit = {
    val text = scala.io.Source.fromFile(args(0)).mkString

    //warmup
    print("warmup\n")
    time {
      val prog = new Program(">++[<+++++++++++++>-]<[[>+>+<<-]>[<+>-]++++++++[>++++++++<-]>[-]<<>++++++++++[>++++++++++[>++++++++++[>++++++++++[>++++++++++[>++++++++++[>++++++++++[-]<-]<-]<-]<-]<-]<-]<-]++++++++++")
      prog.run
    }
    //

    print("run\n")
    time {
      val prog = new Program(text)
      prog.run
    }

  }
}
